src/libstd/panicking.rs

```diff
@@ -195,41 +213,139 @@ fn default_hook(info: &PanicInfo) {
     }
 }
 
-pub fn on_panic(obj: &(Any+Send), file: &'static str, line: u32) {
-    let panics = PANIC_COUNT.with(|s| {
-        let count = s.get() + 1;
-        s.set(count);
-        count
+/// Invoke a closure, capturing the cause of an unwinding panic if one occurs.
+pub unsafe fn try<R, F: FnOnce() -> R>(f: F) -> Result<R, Box<Any + Send>> {
+    let mut slot = None;
+    let mut f = Some(f);
+    let ret = PANIC_COUNT.with(|s| {
+        let prev = s.get();
+        s.set(0);
+
+        let mut to_run = || {
+            slot = Some(f.take().unwrap()());
+        };
+        let fnptr = get_call(&mut to_run);
+        let dataptr = &mut to_run as *mut _ as *mut u8;
+        let r = __rust_maybe_catch_panic(mem::transmute::<_, fn(*mut u8)>(fnptr),
+                                         dataptr);
+        s.set(prev);
+
+        r
     });
 
-    // If this is the third nested call, on_panic triggered the last panic,
-    // otherwise the double-panic check would have aborted the process.
-    // Even if it is likely that on_panic was unable to log the backtrace,
-    // abort immediately to avoid infinite recursion, so that attaching a
-    // debugger provides a useable stacktrace.
-    if panics >= 3 {
+    return ret.map(|()| {
+        slot.take().unwrap()
+    });
+
+    fn get_call<F: FnMut()>(_: &mut F) -> fn(&mut F) {
+        call
```

Comment:
Hm good point! This felt a little weird... The signature of that though is:

```rust
extern "rust-call" fn(&mut F, ());
```

Which I'm slightly hesitant to transmute to `fn(*mut u8)` due to the second argument...